### PR TITLE
L7 deny rule check in authz validation

### PR DESCRIPTION
**Pull request by aryan16**
**Warning message for L7 Deny rule which will block all the tcp traffic under the scope of the policy having that rule**

### CODE REVIEW

**Review by GregHanson — CHANGES_REQUESTED** [review]
Can you add a [release note](https://github.com/istio/istio/tree/master/releasenotes) to highlight the new warning message?

**Comment by aryan16** [comment]
@GregHanson PTAL

File: pkg/config/validation/validation.go [hunk 1]
```diff
@@ -2054,8 +2065,26 @@ var ValidateAuthorizationPolicy = registerValidateFunc("ValidateAuthorizationPol
...
+				(rule.From != nil && len(rule.From) != 0)) && in.Action == security_beta.AuthorizationPolicy_DENY {
+				ns := cfg.Meta.Namespace
+				if in.Selector != nil && in.Selector.MatchLabels != nil {
+					var label string
+					for key, val := range in.Selector.MatchLabels {
+						label = label + " " + key + ":" + val
+					}
+					warning = fmt.Errorf("all the traffic to TCP ports for pods with labels %s in"+
+						"namespace %s will be denied because your deny policy has a pure L7 rule,"+
+						"it's better to add ports when you define any deny rule ", ns, label)
```
Comment (howardjohn) [inline pkg/config/validation/validation.go:2079]: Its a bit confusing that we are looking for "Pure L7". That means this doesn't trigger it:

```
spec:
  action: DENY
  rules:
  - 
    from:
    - source:
        principals: ["cluster.local/ns/default/sa/sleep"]
    to:
    - operation:
        methods: ["GET"]
```

Users intent: "Deny GET from sleep"
Actual impact: "Deny everything from sleep"

This PR would *not* warn about that at all.

File: pkg/config/validation/validation.go [hunk 2]
```diff
@@ -2054,8 +2065,26 @@ var ValidateAuthorizationPolicy = registerValidateFunc("ValidateAuthorizationPol
...
+
+			if !tcpRulesInFrom && !tcpRulesInTo && ((rule.To != nil && len(rule.To) != 0) ||
+				(rule.From != nil && len(rule.From) != 0)) && in.Action == security_beta.AuthorizationPolicy_DENY {
+				ns := cfg.Meta.Namespace
+				if in.Selector != nil && in.Selector.MatchLabels != nil {
+					var label string
+					for key, val := range in.Selector.MatchLabels {
+						label = label + " " + key + ":" + val
+					}
+					warning = fmt.Errorf("all the traffic to TCP ports for pods with labels %s in"+
```
Comment (howardjohn) [inline pkg/config/validation/validation.go:2077]: nit: we are missing space between the string joins.

Comment (howardjohn) [inline pkg/config/validation/validation.go:2077]: nit: I would make warning: "configured AuthorizationPolicy will deny all traffic to TCP ports due to using only HTTP attributes in a DENY rule; it is recommended to explicitly specify the port"

We don't need the labels or namespace, the warning is associated with the resource already

File: pkg/config/validation/validation.go (hunk 1 above)
Comment (aryan16) [reply pkg/config/validation/validation.go:2079]: The attached policy has info for TCP filter as well (`Principals`). The behavior would be, request will be denied if it a GET req or coming from any peer with identity `cluster.local/ns/default/sa/sleep`. IMO it is a valid policy and we shouldn't warn about it.

File: pkg/config/validation/validation.go (hunk 2 above)
Comment (aryan16) [reply pkg/config/validation/validation.go:2077]: sure, thanks

File: pkg/config/validation/validation.go [hunk 3]
```diff
@@ -2054,8 +2069,14 @@ var ValidateAuthorizationPolicy = registerValidateFunc("ValidateAuthorizationPol
 					}
 				}
 			}
+
+			if (fromRuleExist && !toRuleExist && !tcpRulesInFrom) || (toRuleExist && !tcpRulesInTo) {
+				warning = fmt.Errorf("configured AuthorizationPolicy will deny all traffic" +
+					"to TCP ports under its scope due to the use of onlyHTTP attributes in a DENY rule;" +
```
Comment (howardjohn) [inline pkg/config/validation/validation.go:2075]:
```suggestion
					"to TCP ports under its scope due to the use of only HTTP attributes in a DENY rule; " +
```

File: pkg/config/validation/validation.go [hunk 4]
```diff
@@ -2054,8 +2069,14 @@ var ValidateAuthorizationPolicy = registerValidateFunc("ValidateAuthorizationPol
 					}
 				}
 			}
+
+			if (fromRuleExist && !toRuleExist && !tcpRulesInFrom) || (toRuleExist && !tcpRulesInTo) {
+				warning = fmt.Errorf("configured AuthorizationPolicy will deny all traffic" +
```
Comment (howardjohn) [inline pkg/config/validation/validation.go:2074]:
```suggestion
				warning = fmt.Errorf("configured AuthorizationPolicy will deny all traffic " +
```

File: pkg/config/validation/validation_test.go [hunk 5]
```diff
@@ -5537,7 +5544,8 @@ func TestValidateAuthorizationPolicy(t *testing.T) {
 					},
 				},
 			},
-			valid: false,
+			valid:   false,
+			Warning: true,
```
Comment (howardjohn) [inline pkg/config/validation/validation_test.go:5548]: This shouldn't warn? Its not a DENY rule

Comment (aryan16) [reply pkg/config/validation/validation_test.go]: Yeah, sorry I forgot to add the deny condition while raising the warning. Fixed, thanks.

**Review by howardjohn — APPROVED** [review]

**Pull request merged**